cdc/generation.cc

```diff
@@ -300,6 +305,37 @@ future<db_clock::time_point> get_local_streams_timestamp() {
     });
 }
 
+// non-static for testing
+constexpr size_t limit_of_streams_in_topology_description() {
+    // Each stream takes 16B and we don't want to exceed 4MB so we can have
+    // at most 262144 streams but not less than 1 per vnode.
+    return 4 * 1024 * 1024 / 16;
+}
+
+// non-static for testing
+topology_description limit_number_of_streams_if_needed(topology_description&& desc) {
+    int64_t streams_count = 0;
+    for (auto& tr_desc : desc.entries()) {
+        streams_count += tr_desc.streams.size();
+    }
+
+    size_t limit = std::max(limit_of_streams_in_topology_description(), desc.entries().size());
+    if (limit >= streams_count) {
+        return std::move(desc);
+    }
+    size_t vnode_limit = limit / desc.entries().size();
+    auto entries = std::move(desc).entries();
+    auto start = entries.back().token_range_end;
+    for (size_t idx = 0; idx < entries.size(); ++idx) {
+        auto end = entries[idx].token_range_end;
+        if (entries[idx].streams.size() > vnode_limit) {
+            entries[idx].streams = create_stream_ids(idx, start, end, vnode_limit, entries[idx].sharding_ignore_msb);
```

Comment:
I see you pass vnode_limit to a parameter call shard_count. Can you explain why this is correct?

Reply:
This is related to the way token -> stream mapping is calculated.
The calculation takes place in [get_stream function](https://github.com/scylladb/scylla/blob/fd1dd0eac7a303ceaa9f3ff643c4848506b0c85c/cdc/metadata.cc#L44-L45).

```
static cdc::stream_id get_stream(
        const cdc::token_range_description& entry,
        dht::token tok) {
    // The ith stream is the stream for the ith shard.
    auto shard_cnt = entry.streams.size();
    auto shard_id = dht::shard_of(shard_cnt, entry.sharding_ignore_msb, tok);

    if (shard_id >= shard_cnt) {
        on_internal_error(cdc_log, "get_stream: shard_id out of bounds");
    }

    return entry.streams[shard_id];
}
```

As you can see, for a token we compute a shard assuming there are as many shards as streams in that vnode. `entry` is a description of a vnode. `entry.streams` is a collection of streams we've generated for this vnode. `shard_id` is computed using `entry.streams.size()` as shard count.

To make this logic work with trimmed set of streams, we need to create the streams assuming the same shard count as `get_stream` that will be using those streams. If we use different shard counts for those two places, the resulting mapping will be unbalanced.

Does that make sense?

Reply:
I still don't get it. Let me explain how I understand it and you can correct me where I'm wrong.

Each vnode spans between 1 and smp::count shards. So for each vnode, we create between 1 and smp::count streams.

To reduce the number of streams, we can pretend smp::count is smaller. Then each vnode will span a smaller number of shards.

What I'm missing is how a vnode count becomes a shard count.

Reply:
> What I'm missing is how a vnode count becomes a shard count.

vnode count does not become a shard count. `vnode_limit` is a limit of streams a vnode can have.
Maybe the name is not great and it could be renamed to `streams_per_vnode_limit`.

Reply:
This computation of a limit might be confusing:
```
    size_t limit = std::max(limit_of_streams_in_topology_description(), desc.entries().size());
    if (limit >= streams_count) {
        return std::move(desc);
    }
    size_t vnode_limit = limit / desc.entries().size();
```

Normally, `vnode_limit = limit_of_streams_in_topology_description() / desc.entries().size();`.
This `std::max(limit_of_streams_in_topology_description(), desc.entries().size());` is there just to make sure we have at least 1 stream in each vnode.

Reply:
Ah, with the new name I understand it.

Reply:
Please update the patch.

Reply:
@avikivity done